src/library_webgpu.js

```diff
@@ -660,39 +660,50 @@ var LibraryWebGPU = {
     device["pushErrorScope"](WebGPU.ErrorFilter[filter]);
   },
 
+  wgpuDevicePopErrorScope__deps: [
+    '$callUserCallback',
 #if MINIMAL_RUNTIME
-  wgpuDevicePopErrorScope__deps: ['$allocateUTF8'],
+    '$allocateUTF8OnStack',
+#else
+    '$runtimeKeepalivePush', '$runtimeKeepalivePop',
 #endif
+  ],
   wgpuDevicePopErrorScope: function(deviceId, callback, userdata) {
     var device = WebGPU.mgrDevice.get(deviceId);
+    {{{ runtimeKeepalivePush() }}}
     device["popErrorScope"]()["then"](function(gpuError) {
-      if (!gpuError) {
-        {{{ makeDynCall('viii', 'callback') }}}(
-          {{{ gpu.ErrorType.NoError }}}, 0, userdata);
-      } else if (gpuError instanceof GPUOutOfMemoryError) {
-        {{{ makeDynCall('viii', 'callback') }}}(
-          {{{ gpu.ErrorType.OutOfMemory }}}, 0, userdata);
-      } else {
+      {{{ runtimeKeepalivePop() }}}
+      callUserCallback(function() {
+        if (!gpuError) {
+          {{{ makeDynCall('viii', 'callback') }}}({{{ gpu.ErrorType.NoError }}}, 0, userdata);
+        } else if (gpuError instanceof GPUOutOfMemoryError) {
+          {{{ makeDynCall('viii', 'callback') }}}({{{ gpu.ErrorType.OutOfMemory }}}, 0, userdata);
+        } else {
 #if ASSERTIONS
-        assert(gpuError instanceof GPUValidationError);
+          assert(gpuError instanceof GPUValidationError);
 #endif
-        var messagePtr = allocateUTF8(gpuError.message);
-        {{{ makeDynCall('viii', 'callback') }}}({{{ gpu.ErrorType.Validation }}}, messagePtr, userdata);
-        _free(messagePtr);
-      }
+          var messagePtr = allocateUTF8OnStack(gpuError.message);
```

Comment:
Are these usages of allocateUTF8OnStack (from the event loop, not from a WASM call into JS) valid? Convenient if so, as they should avoid a malloc.

Reply:
If you use `allocateUTF8OnStack` (or any function that calls `allocStack` internally like that), I think you also need to wrap with `var stackTop = stackSave();` and `stackRestore(stackTop);`.